File: src/utils/paths.ts
import { homedir } from "node:os";
import { join, resolve } from "node:path";

function home(): string {
	return process.env.HOME ?? homedir();
}

export function graphmindDir(): string {
	return join(home(), ".graphmind");
}

export function configPath(): string {
	return join(graphmindDir(), "config.json");
}

export function memoryDir(): string {
	return join(graphmindDir(), "memory");
}

export function graphsDir(): string {
	return join(graphmindDir(), "graphs");
}

export function crossLinksDir(): string {
	return join(graphmindDir(), "cross-links");
}

export function sessionsDir(): string {
	return join(graphmindDir(), "sessions");
}

export function graphDir(slug: string): string {
	return join(graphsDir(), slug);
}

export function graphDbPath(slug: string): string {
	return join(graphDir(slug), "graph.db");
}

export function memoryPath(slug: string): string {
	return join(memoryDir(), `${slug}.jsonl`);
}

export function globalMemoryPath(): string {
	return join(memoryDir(), "global.jsonl");
}

export function crossLinksPath(): string {
	return join(crossLinksDir(), "links.jsonl");
}

export function sessionLogPath(date: string): string {
	return join(sessionsDir(), `${date}.md`);
}

export function metaPath(slug: string): string {
	return join(graphDir(slug), "meta.json");
}

export function cacheDirPath(slug: string): string {
	return join(graphDir(slug), "cache");
}

export function safePath(path: string, allowedRoots: string[]): string {
	const resolved = resolve(path);
	const isAllowed = allowedRoots.some((root) => resolved.startsWith(resolve(root)));
	if (!isAllowed) {
		throw new Error(`Path traversal blocked: ${path} is outside allowed boundaries`);
	}
	return resolved;
}

File: crates/graphmind-config/src/paths.rs
use std::path::PathBuf;

fn home_dir() -> PathBuf {
    dirs::home_dir().expect("Cannot determine home directory")
}

pub fn graphmind_dir() -> PathBuf {
    home_dir().join(".graphmind")
}

pub fn config_path() -> PathBuf {
    graphmind_dir().join("config.json")
}

pub fn memory_dir() -> PathBuf {
    graphmind_dir().join("memory")
}

pub fn graphs_dir() -> PathBuf {
    graphmind_dir().join("graphs")
}

pub fn cross_links_dir() -> PathBuf {
    graphmind_dir().join("cross-links")
}

pub fn sessions_dir() -> PathBuf {
    graphmind_dir().join("sessions")
}

pub fn graph_dir(slug: &str) -> PathBuf {
    graphs_dir().join(slug)
}

pub fn graph_db_path(slug: &str) -> PathBuf {
    graph_dir(slug).join("graph.db")
}

pub fn cache_dir_path(slug: &str) -> PathBuf {
    graph_dir(slug).join("cache")
}

pub fn memory_path(slug: &str) -> PathBuf {
    memory_dir().join(format!("{slug}.jsonl"))
}

pub fn global_memory_path() -> PathBuf {
    memory_dir().join("global.jsonl")
}

pub fn cross_links_path() -> PathBuf {
    cross_links_dir().join("links.jsonl")
}

pub fn meta_path(slug: &str) -> PathBuf {
    graph_dir(slug).join("meta.json")
}

pub fn embedding_db_path(slug: &str) -> PathBuf {
    graph_dir(slug).join("embeddings.db")
}

pub fn memory_embedding_db_path() -> PathBuf {
    memory_dir().join("memory_embeddings.db")
}
